Add Read Null Terminated String

Brett208 opened the pull request:
Closes #311 

I think tests are fairly robust, although they are not easy to read

Brett208 commented on src/Stream/Reader.h:
```diff
@@ -85,5 +85,26 @@ namespace Stream
+		// Read characters into a string until a null terminator is encountered
+		// Does not include the null terminator in the returned string
+		// Use maxCount to restrict the size of the returned string
+		std::string ReadNullTerminatedString(std::size_t maxCount = SIZE_MAX)
```
I think maxCount is a good idea here, but could discuss removing as well.

DanRStevens replied:
Heh, I did actually stop to think about that for a bit. It makes it seem almost too much like the fixed length field reading code. Though it's not. The stream position here is only advanced up to and consuming the null byte, not up to the maximum length of the read.

I think the parameter is fine, particularly given the default selected.

Brett208 replied:
Thanks for looking it over.

DanRStevens commented on test/Stream/MemoryStreamReader.test.cpp:
```diff
@@ -95,3 +97,36 @@ TEST_F(SimpleMemoryReader, SeekRelativeOutOfBoundsEndPreservesPosition) {
+
+TEST(MemoryReader, ReadNullTerminatedString)
+{
+	std::array<char, 5> buffer{ 'n', 'u', 'l', 'l', '\0' };
```
This could probably be declared as `const`.

DanRStevens commented on test/Stream/MemoryStreamReader.test.cpp:
```diff
@@ -95,3 +97,36 @@ TEST_F(SimpleMemoryReader, SeekRelativeOutOfBoundsEndPreservesPosition) {
+	std::array<char, 5> result;
+
+	// Check null-terminated
+	Stream::MemoryReader reader(&buffer[0], 5);
```
Using `buffer.size()` might be more self documenting that an integer literal.

DanRStevens commented on test/Stream/MemoryStreamReader.test.cpp:
```diff
@@ -95,3 +97,36 @@ TEST_F(SimpleMemoryReader, SeekRelativeOutOfBoundsEndPreservesPosition) {
+	std::copy(str.begin(), str.end(), result.data());
+	
+	// Ensure null-terminator was not copied
+	EXPECT_NE(buffer, result);
```
An `std::string` can contain embedded nulls. As such, you can actually just use a string compare to ensure it stopped before reading the null.

```cpp
assert(std::string("A", 1) != std::string("A\0", 2));
```

I believe the tests can be written without needing to use `result`.

DanRStevens replied:
This check makes an assumption about the last character of the `result` buffer being non-zero, however, the last character is actually left uninitialized, and so could be zero. That's likely the cause of the test failures.

Testing using the `std::string` result directly, rather than copying to a `std::array` may sidestep this problem.

DanRStevens commented on test/Stream/MemoryStreamReader.test.cpp:
```diff
@@ -95,3 +97,36 @@ TEST_F(SimpleMemoryReader, SeekRelativeOutOfBoundsEndPreservesPosition) {
+	buffer[4] = '0';
+	reader.Seek(0);
+	EXPECT_THROW(str = reader.ReadNullTerminatedString(), std::runtime_error);
+	EXPECT_NE(buffer, result);
```
Maybe a comment about where the exception comes from. It wasn't obvious to me at first that this relied on the stream to throw an exception, rather than just reading off into random unbounded memory locations until something (maybe) went wrong.

DanRStevens commented on test/Stream/MemoryStreamReader.test.cpp:
```diff
@@ -95,3 +97,36 @@ TEST_F(SimpleMemoryReader, SeekRelativeOutOfBoundsEndPreservesPosition) {
+	// Check maxCount stops reading string
+	reader.Seek(0);
+	EXPECT_NO_THROW(str = reader.ReadNullTerminatedString(4));
+	EXPECT_EQ(4, str.size());
```
If the string value is tested directly, that also indirectly tests the string length. The test might be a bit more clear about what is happening by testing the string value.

Brett208 commented:
@DanRStevens I think this addresses the points you brought up??? It is definitely better written now.

Looks like I accidentally committed readme changes into this branch...

DanRStevens commented:
I went and split out the ReadMe.md updates into PR #316.

I did a bit of work on the unit tests to hopefully make them a little easier to read, and to test a bit more in terms of stream position expectations.

----

After looking at the unit tests more closely, I wonder if we might change the specification for when the end of the stream has been reached. Maybe we should return the string read so far, rather than throwing an exception? This is of course a bit of an arbitrary choice. One interesting case is it might make sense to throw an exception if you were already at the end of the stream, rather than silently return a string of length 0. Or maybe a string of length 0 is entirely desirable.

If we make a change to not throw, but rather return the string read so far, we could consider using the non-throwing `ReadPartial` method internally.
